### DISCUSSION TITLE
Follow on buffer switch

### DISCUSSION THREAD

**Discussion by Tweekism** [jannis-baum/Vivify#123] [Ideas]
Is there any appetite for an option that makes vivify follow you when you switch buffers?

![_Screencast from 2024-07-25 23-23-02](https://github.com/user-attachments/assets/42d5d925-a188-4c57-88d9-7cb3db88078d)

**Comment by tuurep** [comment]
As an idea: would love this

In practice (implementation): is it a hard problem?

I think one problem related to this would be that different browsers work differently in regards to reusing browser tabs: https://github.com/jannis-baum/Vivify/pull/117#issuecomment-2245916454 onwards

**Reply by jannis-baum** [reply]
Should be quite straight forward. I wouldn't open a new URL from Vim (because as you said this would raise the tab-reusing issue again) but instead create a new API endpoint for switching an existing client to a different page. I.e.

1. vivify.vim sends POST to vivify-server
2. vivify-server sends message to clients at given path through existing websockets
3. client reacts to message by running vanilla client-side JS in the browser that opens the new page on the same tab.

Basically the same way the scrolling and buffer-updating works right now as well!

**Reply by tuurep** [reply]
This sounds really great

Especially for navigating relative file links with `gf` and such

**Reply by jannis-baum** [reply]
Nice! Feel free to implement it ^^

Just as a side note: This way it's all good, but if someone comes up with wanting it the other way around (open new file in Vivify and change Vim buffer as a reaction) it would become a bit of a nightmare :D

**Reply by Tweekism** [reply]
> Should be quite straight forward.

Famous first words, lol 😂

**Reply by Tweekism** [reply]
> Just as a side note: This way it's all good, but if someone comes up with wanting it the other way around (open new file in Vivify and change Vim buffer as a reaction) it would become a bit of a nightmare :D

That sounds like a nightmare in general lol. My expectation would be if you start navigating in the browser it just lets you browse around disconnected from vim.

Would it snap back once you start switching buffers in vim again? My thought was no, but rather would have to do `:Vivify` again to sync it up again or open another tab, something like that.

**Reply by jannis-baum** [reply]
> That sounds like a nightmare in general lol. My expectation would be if you start navigating in the browser it just lets you browse around disconnected from vim.

Yes, exactly ^^ And just to define "nightmare": It would probably require having a second websocket connection between the client (browser tab) and Vim, i.e. Vim would have to have a server running. I don't think I would want this on the Vim plugin ^^ But since Vivify offers adding custom client-side JS someone who really wants it could implement it without having to modify any part of Vivify, it would just need a second Vim plugin or some custom scripting there as well.

> Would it snap back once you start switching buffers in vim again? My thought was no, but rather would have to do :Vivify again to sync it up again or open another tab, something like that.

It would not snap back when you switch buffers in Vim. But if you browse back to the file that's open in Vim and then type something there, it would automatically be synchronized again without having to run anything from Vim.

**Reply by tuurep** [reply]
The remaining big question I think is:

In which cases exactly does this happen? Some kinda on-buffer-switch autocmd, but like, exactly what?

**Reply by jannis-baum** [reply]
I'd guess it should happen on `BufLeave`. Question is if there's a way we can figure out what the next buffer will be from there.

**Reply by tuurep** [reply]
Yeah and do we have to have a sense of "is Vivify currently running for this buffer, or not" ?

**Reply by jannis-baum** [reply]
Not quite but we also don't need it. We can just POST anyways and all clients that exist on the path will update. Or none if there aren't any. That's how the scrolling & auto updating works as well :)

(#103 will introduce a reply that will let you know how many clients there are but it's not necessary for us to read that reply here, no reason to overcomplicate it ^^)

**Reply by twidxuga** [reply]
+1 on this feature. Would love to have Vivify follow me around when I browse my markdown "second brain" in nvim. This would remove the need to open additional browser tabs or explicitly call vivify again. 

In a [previous similar project](https://github.com/twidxuga/vim-instant-markdown), I used BufEnter for markdown typed files to make the switch, triggering a push of the new file's content to the server (though this may not be applicable to vivify).

**Comment by Tweekism** [comment]
Created issue and added to queue

https://github.com/jannis-baum/Vivify/issues/140#issue-2433948842
